### PR TITLE
planner: adjust N used in TopN cost formula based on the total number of rows

### LINKED ISSUE #43285: planner: adjust N used in TopN cost formula based on the total number of rows
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

<!-- a step by step guide for reproducing the bug. -->

Currently, we use the formula `N*row-size*mem-factor` to estimate the memory cost of TopN, but actually, `min(N, total_rows)*row-size*mem-factor` is better.

### 2. What did you expect to see? (Required)

### 3. What did you see instead (Required)

### 4. What is your TiDB version? (Required)

<!-- Paste the output of SELECT tidb_version() -->

```
+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| tidb_version()                                                                                                                                                                                                                                                                                                  |
+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Release Version: v7.1.0-alpha-299-gee4fb2965d
Edition: Community
Git Commit Hash: ee4fb2965d07a6363711ee942d08d29e14bc1eb4
Git Branch: non-prep-debugx
UTC Build Time: 2023-04-20 09:10:06
GoVersion: go1.20.2
Race Enabled: false
TiKV Min Version: 6.2.0-alpha
Check Table Before Drop: false
Store: unistore |
+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
```

### PR DESCRIPTION
**Pull request by qw4990**
### What problem does this PR solve?

Issue Number: close #43285

Problem Summary: planner: adjust N used in TopN cost formula based on the total number of rows

### What is changed and how it works?

planner: adjust N used in TopN cost formula based on the total number of rows

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: planner/core/plan_cost_ver2.go [hunk 1]
```diff
@@ -396,6 +396,7 @@ func (p *PhysicalTopN) getPlanCostVer2(taskType property.TaskType, option *PlanC
 
 	rows := getCardinality(p.children[0], option.CostFlag)
 	n := math.Max(1, float64(p.Count+p.Offset))
+	n = math.Min(n, rows)
```
Comment (hawkingrei) [inline planner/core/plan_cost_ver2.go:399]: In go1.21, we can replace it with ```min```. ```min```/```max``` is a new built-in function.

Comment (qw4990) [reply planner/core/plan_cost_ver2.go:399]: gotit

File: planner/core/integration_test.go [hunk 2]
```diff
@@ -3205,31 +3205,40 @@ func TestLimitPushDown(t *testing.T) {
...
+	tk.MustExec(`insert into t values (1)`)
+	tk.MustExec(`insert into t values (1)`)
+	tk.MustExec(`insert into t values (1)`)
+	tk.MustExec(`insert into t values (1)`)
+	tk.MustExec(`insert into t values (1)`)
+	tk.MustExec(`insert into t values (1)`)
 	tk.MustExec(`analyze table t`)
 
 	tk.MustExec(`set tidb_opt_limit_push_down_threshold=0`)
 	tk.MustQuery(`explain format=brief select a from t order by a desc limit 10`).Check(testkit.Rows(
```
Comment (AilinKid) [inline planner/core/integration_test.go:3220]: better set the LIMIT bigN>10 here?

File: planner/core/integration_test.go [hunk 3]
```diff
@@ -3205,31 +3205,40 @@ func TestLimitPushDown(t *testing.T) {
...
 	tk.MustExec(`insert into t values (1)`)
+	tk.MustExec(`insert into t values (1)`)
+	tk.MustExec(`insert into t values (1)`)
+	tk.MustExec(`insert into t values (1)`)
+	tk.MustExec(`insert into t values (1)`)
+	tk.MustExec(`insert into t values (1)`)
+	tk.MustExec(`insert into t values (1)`)
+	tk.MustExec(`insert into t values (1)`)
+	tk.MustExec(`insert into t values (1)`)
+	tk.MustExec(`insert into t values (1)`)
```
Comment (hawkingrei) [inline planner/core/integration_test.go:3216]: merge those insert statements into one. use batch insert to speed up.

File: planner/core/integration_test.go (hunk 2 above)
Comment (qw4990) [reply planner/core/integration_test.go:3220]: Actually, this test is not for this PR... This test is used to test whether `tidb_opt_limit_push_down_threshold` and `limit_to_cop()` can take effect.

**Review by hawkingrei — APPROVED** [review]

File: planner/core/integration_test.go (hunk 2 above)
Comment (AilinKid) [reply planner/core/integration_test.go:3211]: get

**Pull request merged**